lib/astutils.cpp

```diff
@@ -574,6 +574,10 @@ bool isConstExpression(const Token *tok, const Library& library, bool pure)
     }
     if (tok->tokType() == Token::eIncDecOp)
         return false;
+    if(tok->isAssignmentOp())
+        return false;
+    if(Token::Match(tok, "<<|>>"))
```

Comment:
to me it seems that `1<<2` should be a const expression.

We can use `isLikelyStreamRead` to check if a `>>` is a non-const expression but as far as I remember there is no such function for `<<`.

Reply:
please clarify for me if I am missing something so it should not be a const expression.

Reply:
This was to make sure the FP mentioned in issue 5683 is fixed. Of course, I dont really understand how its a FP, both `0 << 7` and `0 << 0` are the same value.

Reply:
ah.. it is the same value but it's not the same expression.

Reply:
it seems reasonable to skip the warning in this case. It does look like reasonable code. And as far as I can remember, the possible mistake that motivates this checker is copy paste.

Reply:
So the problem is that cppcheck is folding `0 << x` into `0`. This program:

```cpp
void f() {
    return 0 << 7 | 0 << 0;
}
```

Becomes:

```
$ ./bin/cppcheck --enable=style x.cpp --debug-normal
Checking x.cpp ...

##file x.cpp
1: void f ( ) {
2: return 0 | 0 ;
3: }
```

That is why `isSameExpression` is true even when the expressions are different.

Reply:
I see. That should be fixed.

For information; Before we had `ValueFlow` I tried to fold as much as possible because that meant that we could detect more UB. But the cost was that it's not good for various stylistic checks when the original code get lost. Nowadays the goal is to revert my changes and fold as little as possible and rely on `ValueFlow` instead in the checks.

We need to have folding in template declarations. But I do not think that there is any good reason to fold any other types of code, for instance your code.